File: R/ch8-gamma.posterior.interval.R
# ---------------------------------------------------------------------------- #
#                        CHAPTER 8. GAMMA POSTERIOR INTERVALS               ----
# ---------------------------------------------------------------------------- #

# Dataset for the Founta et al. (2018),"Large Scale Crowdsourcing and Characterization of Twitter Abusive Behavior" paper, published in ICWSM 2018. 
# Data is accessed from: https://zenodo.org/records/3678559#.Xk_23y97FhE.
# The dataset provided here includes an hated version of the original dataset, 

# DATA & SETUP
hate.retweets <- read.csv("./Data/hate.retweets.csv",
    header=FALSE)
num.retweets <- rep(NA,nrow(hate.retweets))
for (i in 1:nrow(hate.retweets))  
    num.retweets[i] <- sum(!is.na(hate.retweets[i,]))
y <- num.retweets[num.retweets < 100]
n <- length(y)
# summary(y)
# var(y)


# REMINDER: inverse gamma mean = b/(a-1); inverse gamma variance = b^2/((a-1)^2 * (a-2))
# HYPERPRIOR VALUES
m <- 5; a <- 50; b <- 10


# POSTERIOR PARAMETERS from {G}(9379,1966)
post.a <- a + sum(y)
post.b <- b + n
post.mean <- post.a/post.b
post.var <- post.a/(post.b^2)


# GRAPH OF PRIOR-POSTERIOR AND POSTERIOR QUANTILES
pdf("./Images/poisson.gamma.hate.pdf")
par(oma=c(5,5,2,2), mar=c(0,0,0,0),cex.lab=1.25,mfrow=c(1,2))
ruler <- seq(from=3,to=7,length=500)
plot(ruler,dgamma(ruler,a,b),type="l",lwd=3,col="grey70", ylab="",xlim=c(3,7),ylim=c(0,8))
mtext(side=1,outer=FALSE,expression(paste(theta," Support")),cex=1.5,line=3)
mtext(side=2,outer=TRUE,"Density",cex=1.5,line=3)
lines(ruler,dgamma(ruler,post.a,post.b),lwd=4,col="grey30")
text(4,0.75,"Prior",col="grey70",cex=1.25,adj=0.5)
text(5.72,6,"Posterior",col="grey10",cex=1.25,adj=0.5)
curve1.x <- seq(5.25,7.4,length=20); curve1.y <- 4.50+sin(curve1.x)
lines(curve1.x,curve1.y,lwd=3,lty=2,col="red")
plot(ruler,dgamma(ruler,post.a,post.b),lwd=4,col="grey30",ylab="",xlab="",yaxt="n",
    type="l",xlim=c(4.55,5.0),ylim=c(0,8))
mtext(side=1,outer=FALSE,expression(paste("Posterior ",theta," Support")),cex=1.5,line=3)
g.quantiles <- qgamma(c(0.25,0.5,0.75),post.a,post.b)
for (i in ruler[1:259]) segments(i,0,i,dgamma(i,post.a,post.b), col="grey85",lwd=5)
for (i in ruler[260:264]) segments(i,0,i,dgamma(i,post.a,post.b), col="grey75",lwd=5)
for (i in ruler[264:268]) segments(i,0,i,dgamma(i,post.a,post.b), col="grey50",lwd=5)
for (i in ruler[269:500]) segments(i,0,i,dgamma(i,post.a,post.b), col="grey30",lwd=5)
lines(ruler,dgamma(ruler,post.a,post.b),lwd=5,col="grey30")
abline(h=0,lwd=3)
curve2.x <- seq(4.40,4.70,length=20); curve2.y <- 4.27+cos(curve1.x)
lines(curve2.x,curve2.y,lwd=3,lty=2,col="red")
segments(curve2.x[20],curve2.y[20],curve2.x[20]-0.009,curve2.y[20]+0.30,lwd=3,col="red")
segments(curve2.x[20],curve2.y[20],curve2.x[20]-0.022,curve2.y[20]-0.08,lwd=3,col="red")
dev.off()
system("open   ./Images/poisson.gamma.hate.pdf")
system("pdf2ps ./Images/poisson.gamma.hate.pdf ./Images/poisson.gamma.hate.ps")


# POSTERIOR DENSITY TO THE RIGHT OF 4.75
1-pgamma(4.75,shape=post.a,rate=post.b)


# QUANTILES OF THE POSTERIOR FOR theta
n.sims <- 1000000
theta.large.sample <- rgamma(n.sims,post.a,post.b)
summary(theta.large.sample)


# CREDIBLE INTERVALS FOR theta
vals <- c(0.001,0.01,0.05,0.10)
sort.theta.sample <- sort(theta.large.sample)
round(sort.theta.sample[c(n.sims*vals,n.sims*(1-vals))],3)


# HPD INTERVAL for theta
n.sims <- 10000
mix.dat <- c(rnorm(n.sims/2,1,1), rnorm(n.sims/2,7,2)) 
mix.dens <- density(mix.dat)
alpha <- 0.05
decrement <- 10000
target <- sum(mix.dens$y * mix.dens$x) * alpha
exclude <- sum(mix.dens$y * mix.dens$x)
k <- max(mix.dens$y)
while (exclude > target) {
    k <- k - k/decrement
    exclude <- sum(mix.dens$y[mix.dens$y < k]* mix.dens$x[mix.dens$y < k]) 
    print(paste("exclude: ",exclude," k: ",k))
}


# COMPARING MODELS WITH DIFFERENT HYPERPRIOR VALUES
m1 <- 5; a1 <- 50; b1 <- 10
m2 <- 2; a2 <- 1; b2 <- 1000
post.a1 <- a1 + sum(y); post.b1 <- b1 + n
post.a2 <- a2 + sum(y); post.b2 <- b2 + n
theta1.vals <- rgamma(100000,shape=post.a1,rate=post.b1)
theta2.vals <- rgamma(100000,shape=post.a2,rate=post.b2)
y1 <- y2 <- NULL
for (i in 1:1000)  {
    y1 <- c(y1,rpois(n,sample(theta1.vals,1,replace=TRUE)))
    y2 <- c(y2,rpois(n,sample(theta2.vals,1,replace=TRUE)))
}
y1 <- y1[y1 > 1]; y2 <- y2[y2 > 1]
rbind(summary(y),summary(y1),summary(y2))
 

File: Python/ch8-gamma.posterior.interval.py
"""\
CHAPTER 8. GAMMA POSTERIOR INTERVALS 
"""

# Dataset for the Founta et al. (2018),"Large Scale Crowdsourcing and Characterization of Twitter Abusive Behavior" paper, published in ICWSM 2018. 
# Data is accessed from: https://zenodo.org/records/3678559#.Xk_23y97FhE.

# %%
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import ConnectionPatch
from scipy.stats import gamma, poisson, gaussian_kde

# %%
hate_retweets = pd.read_csv("./data/hate.retweets.csv", header=None)
n = hate_retweets.shape[0]

# %%
num_retweets = hate_retweets.count(axis=1).values
num_retweets = num_retweets[num_retweets < 100]
n = len(num_retweets)
y = num_retweets

# %%
# Hyperprior values
m = 5; a = 50; b = 10

# Posterior parameters
post_a = a + np.sum(y)
post_b = b + n
post_mean = post_a / post_b
post_var = post_a / (post_b**2)

# %%
# POSTERIOR DENSITY TO THE RIGHT OF 4.75
p = 1 - gamma.cdf(4.75, post_a, scale=1/post_b)
print(p)

# %%
fig, axs = plt.subplots(1, 2, figsize=(12, 6))
ruler = np.linspace(3, 7, 500)
axs[0].plot(ruler, gamma.pdf(ruler, a, scale=1/b), 
            color='0.7', linewidth=3, label='Prior')
axs[0].plot(ruler, gamma.pdf(ruler, post_a, 
            scale=1/post_b), color='black', 
            linewidth=3, label='Posterior')
axs[0].text(4, 0.75, "Prior", color='0.7', fontsize=12)
axs[0].text(5, 6, "Posterior", color='0.3', fontsize=12)
axs[0].set_xlabel('$\Theta$  Support')
axs[0].set_ylabel('Density')
axs[0].set_ylim(-0.25, 8.25)
axs[1].plot(ruler, gamma.pdf(ruler, post_a, 
            scale=1/post_b), color='black', linewidth=3)
axs[1].set_xlim(4.55, 5.0)
axs[1].set_ylim(-0.25, 8.25)
axs[1].set_xlabel('Posterior $\Theta$ Support')
axs[1].tick_params(which='both', left=False, 
                   labelleft=False)
axs[1].fill_between(ruler, gamma.pdf(ruler, post_a, 
                     scale=1/post_b), color='0.85')
con = ConnectionPatch(
    xyA=(0.5, 3.55), coordsA=axs[0].get_yaxis_transform(),
    xyB=(0.3, 5), coordsB=axs[1].get_yaxis_transform(), 
    linestyle='dashed', arrowstyle="->", lw=2, 
    color='red', connectionstyle="angle,angleA=-150,angleB=-20,rad=50")
axs[1].add_artist(con)
plt.subplots_adjust(wspace=0)
plt.show()

# %%
# QUANTILES OF THE POSTERIOR FOR theta 
n_sims = 1000000
theta_large_sample = np.random.gamma(post_a, 1/post_b, size=n_sims)
#theta_large_sample = gamma.rvs(post_a, scale=1/post_b, size=n_sims)
pd.DataFrame(theta_large_sample).describe()

# %%
# CREDIBLE INTERVALS FOR theta
vals = np.array([0.001,0.01,0.05,0.10]) 
sorted_theta_sample = np.array(sorted(theta_large_sample)) 
print(np.round(sorted_theta_sample[np.floor(n_sims*vals).astype(int)], 3))
print(np.round(sorted_theta_sample[np.floor(n_sims*(1-vals)).astype(int)], 3))

# %%
# HPD INTERVAL for theta
n_sims = 10000
mix_dat = np.concatenate((np.random.normal(1, 1, int(n_sims/2)), np.random.normal(7, 2, int(n_sims/2))))
mix_dens = gaussian_kde(mix_dat)
x = np.linspace(min(mix_dat), max(mix_dat), n_sims)
y = mix_dens(x)

alpha = 0.05
decrement = 10000
target = np.sum(y * x) * alpha
exclude = np.sum(y * x)
k = max(y)

while exclude > target:
    k -= k / decrement
    mask = y < k
    exclude = np.sum(y[mask] * x[mask])
    print(f"exclude: {exclude}   k: {k}")

# %%
# Model Comparison
m1, a1, b1 = 5, 50, 10
m2, a2, b2 = 2, 1, 1000

post_a1 = a1 + np.sum(y)
post_b1 = b1 + len(y)
post_a2 = a2 + np.sum(y)
post_b2 = b2 + len(y)

theta1_vals = gamma.rvs(a=post_a1, scale=1/post_b1, size=100000)
theta2_vals = gamma.rvs(a=post_a2, scale=1/post_b2, size=100000)

y1 = []
y2 = []
for _ in range(1000):
    y1.extend(poisson.rvs(mu=np.random.choice(theta1_vals, 1, replace=True), size=len(y)))
    y2.extend(poisson.rvs(mu=np.random.choice(theta2_vals, 1, replace=True), size=len(y)))

y1 = np.array(y1)
y2 = np.array(y2)

y1 = y1[y1 > 1]
y2 = y2[y2 > 1]

# Summary of y
print(np.percentile(y, [0, 25, 50, 75, 100]))
# Summary of y1
print(np.percentile(y1, [0, 25, 50, 75, 100]))
# Summary of y2
print(np.percentile(y2, [0, 25, 50, 75, 100]))